File: Java/120_Triangle.java
public class Solution {
    // 9ms
    // 从最下一层到最上一层，动态规划
    // 每一个格子的最小代价是其左右子格子的最小代价加上当前代价
    public int minimumTotal(List<List<Integer>> triangle) {
        if(triangle==null)return 0;
        if(triangle.size()==1)return triangle.get(0).get(0);
        List<Integer> cur = triangle.get(triangle.size()-1), below;
        int c;
        for(int i=triangle.size()-2;i>=0;i--){
            below = cur;
            cur = triangle.get(i);
            for(int j=0;j<cur.size();j++){
                c=cur.get(j);
                cur.set(j, Math.min(below.get(j),below.get(j+1))+c);
            }
        }
        return cur.get(0);
    }
}

File: python/120_Triangle.py
# coding=utf-8
# #problem number 120
class Solution:
    # @param triangle, a list of lists of integers
    # @return an integer
    def minimumTotal(self, triangle):
        if triangle is None or len(triangle)==0:
            return 0
        res=triangle[-1]
        for cur in range(2,len(triangle)+1):
            lastbutone=triangle[-cur]
            now=[]
            for i in range(len(lastbutone)):
                now.append(lastbutone[i]+min(res[i],res[i+1]))
            res=now
        return res[0]
        
triang=[
    [2],
    [3,4],
    [6,5,7],
    [4,1,8,3]]
s = Solution()
print((s.minimumTotal(triang)))